enhancements/ocp-coreos-layering.md

```diff
@@ -0,0 +1,435 @@
+
+The OCP admins at example.corp get a copy of this hotfix RPM into their internal data store, and configure a derived build where the kernel RPMs are applied. The MCO builds a derived image and rolls it out.
+
+(Note: this flow would likely be explained as a customer portal document, etc.)
+
+Later, a fixed kernel with a newer version is released in the main OCP channels.  The override continues to apply, and will currently require manual attention at some point later to remove the override when we can guarantee the non-hotfix kernel has the desired fix.
+
+#### Externally built image
+
+As we move towards having users manage many clusters (10, 100 or more), it will make sense to support building a node image centrally.  This will allow submitting the image to a security scanner or review by a security team before deploying to clusters.
+
+Acme Corp has 300 clusters distributed across their manufacturing centers.  They want to centralize their build system in their main data center, and just distribute those images to single node edge machines.  They provide a `custom-coreos-imagestream` object at installation time, and their node CoreOS image is deployed during the installation of each cluster without a build operation.
+
+In the future, we may additionally support a signature mechanism (IMA/fs-verity) that allows these images can be signed centrally.  Each node verifies this image rather than generating a custom state.
+
+(Note some unanswered questions below)
+
+### API Extensions
+
+We will continue to support MachineConfig.  The exact details of the mechanics of custom builds are still TBD.
+
+### Implementation details
+
+<details>
+
+#### NOTE!  These details are provisional and we intend a forthcoming enhancement to detail more
+    
+4. Inside the cluster, there will be an `openshift-machine-config-operator/coreos` object of type `imagestream`, owned by the MCO.
+5. The MCO builds a container image for each `MachineConfigPool` from user input using the cluster build system 
+7. Each machineconfig pool will also support a `custom-coreos` `BuildConfig` object and imagestream.  This build must use the `mco-coreos` imagestream as a base.   The result of this will be rolled out by the MCO to nodes.
+8. Each machineconfig pool will also support a `custom-external-coreos` imagestream for pulling externally built images (PROVISIONAL)
+9. MCD continues to perform drains and reboots, but writes much less configuration per node
+10. The Machine Configuration Server (MCS) will only serve a "bootstrap" Ignition configuration (pull secret, network configuration) sufficient for the node to pull the target container image.
+
+For clusters without any custom MachineConfig at all, the MCO will deploy the result of the `mco-coreos` build.
+
+<!-- current drawing: https://docs.google.com/drawings/d/1cGly8mjYDezVjUQiUEWd63IO_S2ZrgSbv1cT5X1KJIk/edit -->
```

Comment:
Could we add a section on upgrade flow? It could cover:
1. How we move existing clusters to the new format.
2. How the imagestream is updated as new OpenShift updates go out.
(The information is there in the doc, but it may be easier to follow if we have a dedicated section.)

Reply:
OK, added a bit more about 1.